cmd/rukpakctl/cmd/content.go

```diff
@@ -41,61 +42,51 @@ type options struct {
 	namespace string
 }
 
-var opt options
-
-var contentCmd = &cobra.Command{
-	Use:   "content <bundle name>",
-	Short: "display contents of the specified bundle.",
-	Long:  `display contents of the specified bundle.`,
-	Args: func(cmd *cobra.Command, args []string) error {
-		if len(args) < 1 {
-			return errors.New("requires 2 argument: <bundle name>")
-		}
-		return nil
-	},
-	Run: func(cmd *cobra.Command, args []string) {
-		kubeconfig, err := cmd.Flags().GetString("kubeconfig")
-		if err != nil {
-			fmt.Printf("failed to find kubeconfig location: %+v\n", err)
-			return
-		}
-		// use the current context in kubeconfig
-		config, err := clientcmd.BuildConfigFromFlags("", kubeconfig)
-		if err != nil {
-			fmt.Printf("failed to find kubeconfig location: %+v\n", err)
-			return
-		}
-
-		opt.Client, err = runtimeclient.New(config, runtimeclient.Options{
-			Scheme: scheme,
-		})
-		if err != nil {
-			fmt.Printf("failed to create kubernetes client: %+v\n", err)
-			return
-		}
-
-		if opt.Clientset, err = kubernetes.NewForConfig(config); err != nil {
-			fmt.Printf("failed to create kubernetes client: %+v\n", err)
-			return
-		}
-
-		opt.namespace, err = cmd.Flags().GetString("namespace")
-		if err != nil {
-			opt.namespace = "rukpak-system"
-		}
-
-		err = content(opt, args)
-		if err != nil {
-			fmt.Printf("content command failed: %+v\n", err)
-		}
-	},
-}
-
-func init() {
-	rootCmd.AddCommand(contentCmd)
+func newContentCmd() *cobra.Command {
+	var opt options
+
+	contentCmd := &cobra.Command{
+		Use:   "content <bundle name>",
+		Short: "display contents of the specified bundle.",
+		Long:  `display contents of the specified bundle.`,
+		Args:  cobra.ExactArgs(1),
+		Run: func(cmd *cobra.Command, args []string) {
```

Comment:
I know this is just repurposing the existing cmd implementation but use RunE here instead to avoid os.Exit(1)?

Reply:
I agree this could be improved. `RunE` can be a little awkward to work with as well because it always spits out the usage if an error is returned.

You can disable that with `SilenceUsage`, but that can also be annoying if you have some failures where usage makes sense and some failures where it doesn't.

My general approach is to use a combination of `RunE` and `log.Fatal`. Any errors that are related to validating the CLI args would just be returned. Any runtime errors would get the `log.Fatal` treatment so that the usage is not printed.

Reply:
And if I know cobra can handle all the validation I need, I just use `Run` with `log.Fatal`.